Calculation of Summary statistics for experimental output data as well as Simulation output data

Ati-74 opened the pull request:
The code uploaded here is a single code that can be performed on both experimental and simulated output data to calculate summary statistics.

Ati-74 committed:
CellProfiler Output - Pickle files

Ati-74 committed:
Functionality test

This script helps a user to understand how to work with the main code

Ati-74 committed:
Calculation of density

This core code is written by J.C and A.Y but the changes were implemented to be applicable on experimental(as well as simulation) output data

Ati-74 committed:
calculation of distance vs growth

This code is written by A.Y , the changes have been implemented to make the code applicable on experimental (as well as simulation) output data.

Ati-74 committed:
fitting ellipse

Ati-74 committed:
Calculation of aspect ratio + anisotropy

Ati-74 committed:
class of reading CellProfiler (pickle)output

Ati-74 committed:
Reading pickle files + identifying micro-colonies+ calling SS-related functions

cheekolegend commented on Scripts/Run-SS-On-Exp/scripts/density_calculation.py:
```diff
@@ -0,0 +1,197 @@
+    # Add pixels to img border to avoid cutting off cells
+    add_pixels_to_img_border = 5
+
+    # Initialize storage variables
```
This will not work for CellModeller data; a cellState object does not have the property 'x_center'. The x_center is given by cell.pos[0]. 
If you plan on using the same script for both CellModeller and experimental data, the experimental data should be converted to work with scripts designed to process CellModeller data.

Ati-74 replied:
In the comment below I talked about it completely :)

cheekolegend commented on Scripts/Run-SS-On-Exp/scripts/growth_rate_stats.py:
```diff
@@ -0,0 +1,295 @@
+    if cells_age_filtered.shape[0] > 0:
+
+        # Get single-cell properties
+        growth_rates = cells_age_filtered['growth_rate'].values.tolist()
```
This will not work with CellModeller data since it isn't in dataFrame format.

Ati-74 replied:
In the comments below I talked about it completely :)

cheekolegend commented on Scripts/Run-SS-On-Exp/scripts/lownerJohnEllipse.py:
```diff
@@ -0,0 +1,393 @@
+# reference:
```
The license agreement for this script must be included.

cheekolegend commented on Scripts/Run-SS-On-Exp/scripts/summary_statistics.py:
```diff
@@ -0,0 +1,156 @@
+    return x_co_vertex, y_co_vertex
+
+
+def fit_ellipse(bac_in_micro_colony):
```
Add documentation to this function.

cheekolegend commented on Scripts/Run-SS-On-Exp/scripts/summary_statistics.py:
```diff
@@ -0,0 +1,156 @@
+
+
+# find co vertexes
+def find_co_vertex(center_x_list, center_y_list, minor_list, angle_rotation_list):
```
Add documentation to this function.

cheekolegend commented on Scripts/Run-SS-On-Exp/scripts/summary_statistics.py:
```diff
@@ -0,0 +1,156 @@
+    return aspect_ratio
+
+
+def anisotropy_calc(bac_in_micro_colony, neighbour_distance):
```
Add documentation to this function.

cheekolegend commented on Scripts/Run-SS-On-Exp/scripts/MicroColonyAnalysis.py:
```diff
@@ -0,0 +1,322 @@
+            setattr(self, key, my_dict[key])
+
+
+def create_pickle_files(micro_colonies_in_current_time_step, local_aspect_ratio_list, local_anisotropy_list,
```
Add documentation to this function.

Ati-74 replied:
done!

cheekolegend commented on Scripts/Run-SS-On-Exp/scripts/MicroColonyAnalysis.py:
```diff
@@ -0,0 +1,322 @@
+    write_to_pickle_file(data, path, time_step, num_files)
+
+
+def write_to_pickle_file(data, path, time_step, num_files):
```
Add documentation to this function.

Ati-74 replied:
all done!

cheekolegend commented on Scripts/Run-SS-On-Exp/scripts/MicroColonyAnalysis.py:
```diff
@@ -0,0 +1,322 @@
+    return adjacency_matrix
+
+
+def func_current_bacteria_info(cs, dt):
```
Add documentation to this function. The function name should be more specific. Something like convert_cellstates_to_dataframe would be more appropriate. 

I thought we agreed that the experimental data was going to be converted to CellModeller format (i.e. CellState objects), rather than convert CellModeller data to experimental format. Let's review this point in a meeting.

Ati-74 replied:
documentation is added.
For the 2nd part: the input of the script is "pickle file" (this is done by the script I uploaded in the image processing repo), and this script (uploaded here) changes the pickles to data frames for timing efficiency (I'd say this is a part that may confuse you) 
So this means we are on the same page and we can discuss different reasons for doing this (i.e. time required for finding the adjacency matrix would be really too much if I didn't use a data frame) in the meeting.

cheekolegend commented on Scripts/Run-SS-On-Exp/scripts/MicroColonyAnalysis.py:
```diff
@@ -0,0 +1,322 @@
+    # number of digits of last pickle file number
+    num_digit_last_pickle_file_num = len(str(len(filename_list)))
+
+    # In order to identify merged micro colonies, micro colony labels are stored.
```
Identifying separate microcolonies should be optional to the user. This will waste resources when processing thousands of CellModeller simulations. Also, the experimental data may already be pre-processed to include only a single colony in a series of images, so finding merging of microcolonies would be extra overhead.

Ati-74 replied:
all done! now the script is working on both user-preferred modes : local -global
unit test for a user convenience is provided.

cheekolegend commented on Scripts/Run-SS-On-Exp/scripts/MicroColonyAnalysis.py:
```diff
@@ -0,0 +1,322 @@
+    """
+    Goal: this is the main function; the function calls other function that are described above and
+    calculates aspect ratio for each micro colony in each time-step and store the outputs in png format and pickle.
+
```
Include documentation on function arguments.

Ati-74 replied:
all done!

cheekolegend reviewed (changes requested):
We agreed in a meeting with Brian that the experimental data should be processed to be in CellModeller format if we are going to use the scripts to calculate summary statistics for both experiments/simulations. The reasoning was that we need to process simulation data thousands of times, whereas experimental data is only processed once. Your scripts require the opposite: they require simulation data to be in experimental data format. Let's discuss in a meeting before moving forward.

Two ways forward: 
1. We merge these commits, but only use these scripts only for experimental data, or,
2. Write a script that converts experimental data to CellModeller format (as was originally discussed) such that any scripts written for CellModeller do not require modifications for processing experimental data.

Regarding the CP output pickle files, only upload enough for a demo. Two or three files will do.

Every function written needs to have a doc string describing what it does, what the arguments are, and what it returns.

Ati-74 committed:
Create .gitignore

Ati-74 committed:
CellModeller output pickle files

Ati-74 committed:
CellProfiler output pickle files

Ati-74 committed:
files have been removed (because I changed the name of the directory since it wasn't a good representative of the whole capability of that)

Ati-74 committed:
Functionality test for CellModeller AS WELL AS CellProfiler output pickle files

It shows how we can work with the scripts

Ati-74 committed:
fit ellipse

Ati-74 committed:
file has been removed

Ati-74 committed:
fit ellipse

Ati-74 committed:
this is a single code that is capable of calculating summary statistics of both exp. and sim.

For both "LOCAL" and "Global" modes that can be chosen by a user!

Ati-74 committed:
calculation of summary statistics over time lapse has been added

Documentation describing functions are added.

Ati-74 committed:
this function is used to read experimental output pickle files

Ati-74 committed:
density calculation

Written by J.C and A.Y

Ati-74 committed:
Calculation of dis vs growth rate

Written by A.Y

Ati-74 committed:
calculation of summary statistics over time lapse

Documentation describing functions are added.

Ati-74 committed:
file has been removed

Ati-74 committed:
documentation describing functions are added

Ati-74 committed:
Update .gitignore

Ati-74 committed:
default values are defined for the following parameters:

"mode: global" + "max-distance-between-cells=3.4" + "um-pixel-ratio=0.144"+"min-size-of-microcolony=2"
also in global mode: calculation of anisotropy and aspect ratio is done when the min num of bacteria in the t point is at least 2.

Ati-74 committed:
location of some parameters have been changed in the code

note: the point is if a parameter should be defined with a default value, we should define it at the end of the the input arguments.

Ati-74 committed:
Update unit_test.py

Ati-74 committed:
Create .gitignore
